src/exosite.c

```diff
@@ -251,118 +257,162 @@ uint8_t exo_is_op_write(exo_op *op)
+              new_seq = (new_seq << (8*j)) | opt.val[j];
+            }
+
+            payload = coap_get_payload(&pdu);
+            if (payload.len == 0) {
+              op[i].value[0] = '\0';
+            } else if (payload.len+1 > op[i].value_max || op[i].value == 0) {
+              op[i].state = EXO_REQUEST_ERROR;
+            } else{
+              memcpy(op[i].value, payload.val, payload.len);
+              op[i].value[payload.len] = 0;
+              op[i].mid = coap_get_mid(&pdu);
+              // TODO: User proper logic to ensure it's a new value not a different, but old one.
+              if (op[i].obs_seq != new_seq) {
+                op[i].state = EXO_REQUEST_SUB_ACK_NEW;
+                op[i].obs_seq = new_seq;
               } else {
-                op[i].state = EXO_REQUEST_ERROR;
+                op[i].state = EXO_REQUEST_SUB_ACK;
               }
-              break;
-              break;
-            default:
-              break;
-          }
-          device_state = EXO_STATE_GOOD;
-        } else {
-          op[i].state = EXO_REQUEST_ERROR;
-          if (coap_get_code(&pdu) == CC_UNAUTHORIZED){
-            device_state = EXO_STATE_BAD_CIK;
-            op[i].timeout = now + 10000000;
+            }
+          } else if (coap_get_code_class(&pdu) != 2) {
+            op[i].state = EXO_REQUEST_ERROR;
           }
+          break;
         }
-        break;
-      }else if (op[i].state == EXO_REQUEST_SUBSCRIBED && coap_get_token(&pdu) == op[i].token) {
+      } else if (coap_get_type(&pdu) == CT_ACK) {
+        if (op[i].state == EXO_REQUEST_PENDING && op[i].mid == coap_get_mid(&pdu)) {
+          if (coap_get_code_class(&pdu) == 2) {
+            switch (op[i].type) {
+              case EXO_WRITE:
+                op[i].state = EXO_REQUEST_SUCCESS;
+                break;
+              case EXO_READ:
+                payload = coap_get_payload(&pdu);
+                if (payload.len == 0) {
+                  op[i].value[0] = '\0';
+                } else if (payload.len+1 > op[i].value_max || op[i].value == 0) {
+                  op[i].state = EXO_REQUEST_ERROR;
+                } else{
+                  memcpy(op[i].value, payload.val, payload.len);
+                  op[i].value[payload.len] = 0;
+                  op[i].state = EXO_REQUEST_SUCCESS;
+                }
+                break;
+              case EXO_SUBSCRIBE:
+                payload = coap_get_payload(&pdu);
+                if (payload.len == 0) {
+                  op[i].value[0] = '\0';
+                } else if (payload.len+1 > op[i].value_max || op[i].value == 0) {
+                  op[i].state = EXO_REQUEST_ERROR;
+                } else{
+                  memcpy(op[i].value, payload.val, payload.len);
+                  op[i].value[payload.len] = 0;
+                  op[i].state = EXO_REQUEST_SUCCESS;
+                  op[i].timeout = exopal_get_time() + 120000000 + (rand() % 15 * 100000);
```

Comment:
I think a comment here would suffice, just saying what is going on.  Is `120000000` seconds?    Why is the rand()%15 being scaled by `100000`?  Can you add parenthesis in `rand() % 15 * 100000`?  It would make the order of operation, and likely the intent, more clear.